src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, EditorCompletion.CompletionFilter>();
+
+            var itemsBuilder = ImmutableArray.CreateBuilder<EditorCompletion.CompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+                item.Properties.AddProperty(TriggerBuffer, triggerLocation.Snapshot.TextBuffer);
+                if (!string.IsNullOrEmpty(item.DisplayText))
+                {
+                    itemsBuilder.Add(item);
+                }
+            }
+
+            var items = itemsBuilder.ToImmutable();
+
+            var suggestionItemOptions = completionList.SuggestionModeItem != null
+                    ? new EditorCompletion.SuggestionItemOptions(
+                        completionList.SuggestionModeItem.DisplayText,
+                        (completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                            ? description
+                            : string.Empty))
+                    : null;
+
+            session.Properties.AddProperty(MustSetSelection, suggestionItemOptions != null);
+
+            return new EditorCompletion.CompletionContext(
+                items,
+                suggestionItemOptions,
+                suggestionItemOptions == null ? EditorCompletion.InitialSelectionHint.RegularSelection : EditorCompletion.InitialSelectionHint.SoftSelection);
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return null;
+            }
+
+            var document = triggerBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
```

Comment:
this cast is unsafe.  There is no guarantee a specific language will use CompletionServiceWithProviders.  Furthermore, CompletionService already has GetDescriptionAsync... so why are we not calling htat?

Reply:
Great! Thank you!

---
In reply to: [222082753](https://github.com/dotnet/roslyn/pull/29016#discussion_r222082753) [](ancestors = 222082753)